### PR TITLE
Support for writing ESMRY file 

### PR DESCRIPTION
**Pull request by tskille**
adding support for writing lodsmry
 - ESMRY file is re-written for every time step
 - file ESRMY optimized for fast loading of selected vectors
 - This PR is dependent on PR OPM/opm-simulators#3453

### CODE REVIEW

**Comment by joakim-hove** [comment]
I have not looked at this yet - and we will eventually merge it; but my first comment is the name "lodsmry" - I don't like the "lod" name, and don't think the load on demand semantics is interesting for the user. I don't have a good alternative, but maybe something with "hdf" - or alternative something with e.g. "fast"?

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by tskille** [comment]
This PR should now be ready for review. 

This has been tested on Equinor's linux cluster in Stavanger. A full field model with 15000 summary vectors and approximately 1350 time steps has been run with MPI (4 processors ). 

Simulations shows no significant increase in runtime when using this option. The format defined with LODSMRY file shows dramatic improvement in efficiency when loading individual summary vectors from disk.     

example loading 6 summary vectors from 4 runs (total of 24 vectors)
```
 SMSPEC/UNSMRY:  10.5 sec
 LODSMRY : 0.34 sec
```

**Comment by bska** [comment]
Hm, there appears to be a merge conflict in [src/opm/output/eclipse/EclipseIO.cpp](https://github.com/OPM/opm-common/blob/e7192c69abb15e6b0a63cf3372e24c448afce72f/src/opm/output/eclipse/EclipseIO.cpp#L251-L252), probably as a result of #2596 that I merged earlier today.  I **believe** rebasing this onto current master should be fairly easy to do and that the conflict is just in the `Sumary::add_timestep()` call.

**Comment by tskille** [comment]
Thanks @bska, the branch has now been re-based with master.

**Comment by bska** [comment]
jenkins build this opm-simulators=3453 please

File: src/opm/io/eclipse/LODSmryOutput.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        throw std::invalid_argument("global index out of range");
+
+    std::array<int, 3> result;
+    result[2] = globInd / (dims[0] * dims[1]);
+
+    int rest = globInd % (dims[0] * dims[1]);
+
+    result[1] = rest / dims[0];
+    result[0] = rest % dims[0];
+
```
Comment (bska) [inline src/opm/io/eclipse/LODSmryOutput.cpp:181]: If you write this in the reverse order you get the same result with fewer arithmetic operations
```C++
result[0] = globInd % dims[0];  globInd /= dims[0];
result[1] = globInd % dims[1];
result[2] = globInd / dims[1];
```

File: src/opm/io/eclipse/LODSmryOutput.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    for (size_t n = 0; n < static_cast<size_t>(m_nVect); n++)
+        m_smrydata.push_back({});
+}
+
+
+void LODSmryOutput::write(const std::vector<float>& ts_data, int report_step)
+{
+
+    if (ts_data.size() != static_cast<size_t>(m_nVect))
+        throw std::invalid_argument("size of ts_data vector not same as number of smry vectors");
```
Comment (bska) [inline src/opm/io/eclipse/LODSmryOutput.cpp:79]: You should probably include `<stdexcept>` to get a declaration of `std::invalid_argument`.

**Review by bska — COMMENTED** [review]
This looks good to me.  I just have a couple of minor comments.  I'm also not entirely sure we need the new `basename` parameter in the `EclipseIO` constructor, but I'll defer that to a later time.

**Comment by tskille** [comment]
Thanks for your comments. I believe that I have included all of these in the updated commit.

**Review by bska — APPROVED** [review]
> I believe that I have included all of these in the updated commit.

Thanks for the updates.  This does indeed look good to me.  I'll run one more build check and then merge this if that comes back green.

**Comment by bska** [comment]
jenkins build this opm-simulators=3453 please

**Comment by magnesj** [comment]
> I have not looked at this yet - and we will eventually merge it; but my first comment is the name "lodsmry" - I don't like the "lod" name, and don't think the load on demand semantics is interesting for the user. I don't have a good alternative, but maybe something with "hdf" - or alternative something with e.g. "fast"?

@joakim-hove See more comments related to naming of these files here https://github.com/OPM/opm-simulators/pull/3453

**Comment by tskille** [comment]
> and don't think the load on demand semantics is interesting for the user

Notice that Fetch On Demand is used by the commercial application S3Graf (from Sciencesoft), and this concept has been around from a long time. This is where I have gotten the idea(inspiration) of LOD from. S3Graf has been used by a large number of engineers.

**Comment by tskille** [comment]
@bska PR is updated with changes agreed on in https://github.com/OPM/opm-simulators/pull/3453.

**Comment by bska** [comment]
jenkins build this opm-simulators=3453 please

**Review by bska — APPROVED** [review]
Thanks for the updates.  This looks good to me now and I'll merge into master.

**Pull request merged**